rcl_action/src/rcl_action/action_client.c

```diff
@@ -0,0 +1,812 @@
+{
+  assert(NULL != feedback_subscription);
+  assert(NULL != node);
+  assert(NULL != feedback_message_type_support);
+  assert(NULL != action_name);
+  assert(NULL != feedback_subscription_options);
+
+  rcl_ret_t ret;
+  rcl_allocator_t allocator = feedback_subscription_options->allocator;
+
+  char * feedback_topic_name = NULL;
+  ret = rcl_action_get_feedback_topic_name(action_name, allocator, &feedback_topic_name);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_ERROR;
+  }
+
+  *feedback_subscription = rcl_get_zero_initialized_subscription();
+
+  ret = rcl_subscription_init(
+    feedback_subscription, node,
+    feedback_message_type_support,
+    feedback_topic_name,
+    feedback_subscription_options);
+
+  allocator.deallocate(feedback_topic_name, allocator.state);
+
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_TOPIC_NAME_INVALID == ret) {
+      return RCL_RET_ACTION_NAME_INVALID;
+    }
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_ERROR;
+  }
+  return RCL_RET_OK;
+}
+
+// \internal Initialize subscription to the given action's status topic.
+static rcl_ret_t rcl_action_status_subscription_init(
+  rcl_subscription_t * status_subscription, const rcl_node_t * node,
+  const rosidl_message_type_support_t * status_message_type_support,
+  const char * action_name,
+  const rcl_subscription_options_t * status_subscription_options)
+{
+  assert(NULL != status_subscription);
+  assert(NULL != node);
+  assert(NULL != status_message_type_support);
+  assert(NULL != action_name);
+  assert(NULL != status_subscription_options);
+
+  rcl_ret_t ret;
+  rcl_allocator_t allocator = status_subscription_options->allocator;
+
+  char * status_topic_name = NULL;
+  ret = rcl_action_get_status_topic_name(action_name, allocator, &status_topic_name);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_ERROR;
+  }
+
+  *status_subscription = rcl_get_zero_initialized_subscription();
+
+  ret = rcl_subscription_init(
+    status_subscription, node,
+    status_message_type_support,
+    status_topic_name,
+    status_subscription_options);
+
+  allocator.deallocate(status_topic_name, allocator.state);
+
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_TOPIC_NAME_INVALID == ret) {
+      return RCL_RET_ACTION_NAME_INVALID;
+    }
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_ERROR;
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_client_init(
+  rcl_action_client_t * action_client,
+  rcl_node_t * node,
+  const rosidl_action_type_support_t * type_support,
+  const char * action_name,
+  const rcl_action_client_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_client, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
```

Comment:
`rcl_node_is_valid()` already checks for a null pointer, so this line can be removed.

Reply:
Removed in 0c313821659b90f03650dfcefcbff453acba82f6.